tool/tsh/aws.go

```diff
@@ -0,0 +1,344 @@
+	if len(m) == 0 {
+		return
+	}
+	t := asciitable.MakeTable([]string{columnName})
+	for k := range m {
+		t.AddRow([]string{k})
+	}
+	fmt.Println(t.AsBuffer().String())
+
+}
+
+// findARNBasedOnRoleName tries to match roleName parameter with allowed user ARNs obtained from the Teleport API based on
+// user roles profile. If there is a match the IAM role is created based on accountID and roleName fields.
+func findARNBasedOnRoleName(cf *CLIConf, tc *client.TeleportClient, profile *client.ProfileStatus, accountID, roleName string) (string, error) {
+	validRolesName := make(map[string]struct{})
+	for _, v := range profile.Roles {
+		role, err := tc.GetRole(cf.Context, v)
+		if err != nil {
+			return "", trace.Wrap(err)
+		}
+		for _, v := range role.GetAWSRoleARNs(types.Allow) {
+			arn, err := awsarn.Parse(v)
+			if err != nil {
+				return "", trace.Wrap(err)
+			}
+
+			// Example of the ANR Resource: 'role/EC2FullAccess'
+			parts := strings.Split(arn.Resource, "/")
+			if len(parts) != 2 {
+				continue
+			}
+			if parts[0] != "role" {
+				continue
+			}
+
+			if arn.AccountID == accountID {
+				validRolesName[parts[1]] = struct{}{}
+			}
+			if arn.AccountID == accountID && parts[1] == roleName {
+				return arn.String(), nil
+			}
+		}
+	}
+	if len(validRolesName) != 0 {
+		printMapKeysAs(validRolesName, "Available Roles")
+	}
+	return "", trace.NotFound("failed to find role ARN based on AWSAccountID(%q) and RoleName(%q)", accountID, roleName)
+}
+
+func setFakeAWSEnvCredentials(keyID, accessKey string) error {
+	if err := os.Setenv("AWS_ACCESS_KEY_ID", keyID); err != nil {
+		return trace.Wrap(err)
+	}
+	if err := os.Setenv("AWS_SECRET_ACCESS_KEY", accessKey); err != nil {
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+func getARNFromFlags(cf *CLIConf, tc *client.TeleportClient, profile *client.ProfileStatus, app types.Application) (string, error) {
+	if cf.AWSRoleARN == "" && cf.AWSRoleName == "" {
+		return "", trace.BadParameter("either --aws-role-arn or --aws-role-name flag is required")
+	}
+	if cf.AWSRoleARN != "" {
+		if err := validateARNRole(cf, tc, profile, cf.AWSRoleARN); err != nil {
+			return "", trace.Wrap(err)
+		}
+		return cf.AWSRoleARN, nil
+	}
+	// Try to construct ARN value based on RoleName and APP AWSAccountID.
+	accountID, ok := app.GetAllLabels()[constants.AWSAccountIDLabel]
+	if !ok {
+		// APP configuration doesn't contain an accountID value.
+		return "", trace.BadParameter("the role name is ambiguous, please provide role ARN by --aws-role-arn flag")
+	}
+	var err error
+	arn, err := findARNBasedOnRoleName(cf, tc, profile, accountID, cf.AWSRoleName)
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+	return arn, nil
+}
+
+type tempSelfSignedLocalCert struct {
+	cert   tls.Certificate
+	caFile *os.File
+}
+
+func newTempSelfSignedLocalCert() (*tempSelfSignedLocalCert, error) {
+	caKey, caCert, err := tlsca.GenerateSelfSignedCA(pkix.Name{
+		CommonName:   "AWS Local Proxy",
```

Comment:
Would it make sense to set CN to `localhost` also so if some other client connects and wants to validate hostname using CN?

Reply:
Since go1.15 the behaviour where hostName is verified base on  common name is SAN is not provided is dedicated but if we want to support other client it makes sense.